File: examples/python/channel_example.py
#!/usr/bin/env python3
"""
Example: Using IPC Channel for Message Passing

This example demonstrates the high-level IpcChannel API for
structured message passing between processes.
"""

import time
import threading
from ipckit import IpcChannel


def server(channel_name: str):
    """Server that handles JSON messages."""
    print(f"[Server] Creating channel: {channel_name}")
    channel = IpcChannel.create(channel_name)
    print("[Server] Waiting for client...")

    channel.wait_for_client()
    print("[Server] Client connected!")

    # Handle messages
    while True:
        data = channel.recv_json()
        print(f"[Server] Received: {data}")

        if data.get("type") == "exit":
            channel.send_json({"status": "goodbye"})
            break

        # Process and respond
        if data.get("type") == "ping":
            channel.send_json({"type": "pong", "id": data.get("id")})
        elif data.get("type") == "compute":
            result = sum(data.get("numbers", []))
            channel.send_json({"type": "result", "value": result})
        else:
            channel.send_json({"type": "error", "message": "Unknown type"})

    print("[Server] Done!")


def client(channel_name: str):
    """Client that sends JSON messages."""
    time.sleep(0.5)  # Wait for server

    print(f"[Client] Connecting to channel: {channel_name}")
    channel = IpcChannel.connect(channel_name)
    print("[Client] Connected!")

    # Send ping
    channel.send_json({"type": "ping", "id": 1})
    response = channel.recv_json()
    print(f"[Client] Ping response: {response}")

    # Send compute request
    channel.send_json({"type": "compute", "numbers": [1, 2, 3, 4, 5]})
    response = channel.recv_json()
    print(f"[Client] Compute response: {response}")

    # Send another ping
    channel.send_json({"type": "ping", "id": 2})
    response = channel.recv_json()
    print(f"[Client] Ping response: {response}")

    # Exit
    channel.send_json({"type": "exit"})
    response = channel.recv_json()
    print(f"[Client] Exit response: {response}")

    print("[Client] Done!")


def main():
    channel_name = f"example_channel_{time.time_ns()}"

    # Run server and client in separate threads
    server_thread = threading.Thread(target=server, args=(channel_name,))
    client_thread = threading.Thread(target=client, args=(channel_name,))

    server_thread.start()
    client_thread.start()

    server_thread.join()
    client_thread.join()

    print("\nDone!")


if __name__ == "__main__":
    main()

File: examples/rust/channel_example.rs
//! Example: Using IPC Channel for Message Passing in Rust
//!
//! Run with: cargo run --example channel_example

use ipckit::{IpcChannel, Result};
use serde::{Deserialize, Serialize};
use std::thread;
use std::time::Duration;

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
enum Message {
    #[serde(rename = "ping")]
    Ping { id: u32 },
    #[serde(rename = "pong")]
    Pong { id: u32 },
    #[serde(rename = "compute")]
    Compute { numbers: Vec<i32> },
    #[serde(rename = "result")]
    Result { value: i32 },
    #[serde(rename = "exit")]
    Exit,
    #[serde(rename = "goodbye")]
    Goodbye,
}

fn main() -> Result<()> {
    let channel_name = format!("example_channel_{}", std::process::id());

    // Spawn server thread
    let server_channel_name = channel_name.clone();
    let server_handle = thread::spawn(move || -> Result<()> {
        println!("[Server] Creating channel: {}", server_channel_name);
        let mut channel = IpcChannel::<Message>::create(&server_channel_name)?;
        println!("[Server] Waiting for client...");

        channel.wait_for_client()?;
        println!("[Server] Client connected!");

        loop {
            let msg: Message = channel.recv()?;
            println!("[Server] Received: {:?}", msg);

            match msg {
                Message::Ping { id } => {
                    channel.send(&Message::Pong { id })?;
                }
                Message::Compute { numbers } => {
                    let value = numbers.iter().sum();
                    channel.send(&Message::Result { value })?;
                }
                Message::Exit => {
                    channel.send(&Message::Goodbye)?;
                    break;
                }
                _ => {}
            }
        }

        println!("[Server] Done!");
        Ok(())
    });

    // Wait for server to start
    thread::sleep(Duration::from_millis(500));

    // Client
    println!("[Client] Connecting to channel: {}", channel_name);
    let mut channel = IpcChannel::<Message>::connect(&channel_name)?;
    println!("[Client] Connected!");

    // Send ping
    channel.send(&Message::Ping { id: 1 })?;
    let response: Message = channel.recv()?;
    println!("[Client] Response: {:?}", response);

    // Send compute
    channel.send(&Message::Compute {
        numbers: vec![1, 2, 3, 4, 5],
    })?;
    let response: Message = channel.recv()?;
    println!("[Client] Response: {:?}", response);

    // Exit
    channel.send(&Message::Exit)?;
    let response: Message = channel.recv()?;
    println!("[Client] Response: {:?}", response);

    server_handle.join().unwrap()?;
    println!("\nDone!");

    Ok(())
}
